### PR TITLE
Inviscid Cp Residual: predict viscous correction instead of raw p

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model currently predicts raw pressure from geometry+flow conditions. By computing an analytical inviscid pressure baseline (thin-airfoil Cp) and changing the training target from p to p_residual = p - p_inviscid, the model only needs to learn the viscous correction. This correction is:
- **Smaller in magnitude** — easier to learn
- **More directly correlated with Re** — viscous effects scale with Reynolds number
- **Smoother** — viscous corrections vary slowly compared to raw pressure (which has leading-edge singularities)

**Key distinction from thorfinn's panel-Cp feature (#2319):** That PR feeds inviscid Cp as an INPUT feature (additive signal). This PR changes the TRAINING TARGET — a fundamentally different approach. Both could work independently.

**Reference:** B-GNN (arXiv:2503.18638, Zhang et al., March 2025) reports 88% OOD error reduction using inviscid Cp + residual prediction in a similar aerodynamics setting.

### Step 1: Add arguments

```python
parser.add_argument('--inviscid_residual', action='store_true',
                    help='Predict viscous residual (p - p_inviscid) instead of raw p')
```

### Step 2: Compute inviscid Cp approximation

Add this function (thin-airfoil theory, vectorized):

```python
def compute_inviscid_cp(x_feat, surf_mask):
    """Compute thin-airfoil inviscid Cp for surface nodes. Returns [B, N, 1]."""
    B, N, _ = x_feat.shape
    cp_inv = torch.zeros(B, N, 1, device=x_feat.device, dtype=x_feat.dtype)
    
    for b in range(B):
        mask = surf_mask[b]
        if mask.sum() < 2:
            continue
        # Arc-length normalized position (0=LE, 1=TE)
        # Check which column saf_norm is — typically x_feat[:, :, 8]
        arc = x_feat[b, mask, 8]  # adjust column index if needed
        arc_norm = (arc - arc.min()) / (arc.max() - arc.min() + 1e-8)
        
        # AoA in radians — check column index
        aoa = x_feat[b, mask, 14:15].mean()  # global AoA feature, adjust column
        
        # Thin airfoil Cp: Cp = -2*sin(alpha) / sqrt(xi*(1-xi))
        xi = arc_norm.clamp(0.02, 0.98)  # avoid singularities at LE/TE
        denom = torch.sqrt(xi * (1.0 - xi)).clamp(min=0.01)
        
        # Upper/lower surface sign (check which feature indicates this)
        y = x_feat[b, mask, 1]  # y-coordinate
        side = torch.sign(y - y.mean())
        
        cp_inv[b, mask, 0] = -side * 2.0 * torch.sin(aoa.abs()) / denom
        cp_inv[b, mask, 0] = cp_inv[b, mask, 0].clamp(-4.0, 2.0)
    
    return cp_inv
```

### Step 3: Modify training target

In the training loop, BEFORE computing loss:

```python
if args.inviscid_residual:
    cp_inviscid = compute_inviscid_cp(x, surf_mask)
    # Only modify pressure channel (index 2) for surface nodes
    target_modified = target.clone()
    target_modified[:, :, 2:3] = target[:, :, 2:3] - cp_inviscid
    # Use target_modified for loss computation
    # At validation, add back: pred_p = model_output[:, :, 2:3] + cp_inviscid
```

### Step 4: Modify validation to add back inviscid Cp

In the validation loop, after model prediction:

```python
if args.inviscid_residual:
    cp_inviscid_val = compute_inviscid_cp(x_val, surf_mask_val)
    pred_val[:, :, 2:3] = pred_val[:, :, 2:3] + cp_inviscid_val
    # Now pred_val contains the full pressure (not residual)
    # The val metric computation on full pressure is apples-to-apples with baseline
```

### Step 5: Run

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/inviscid-residual-s42" --wandb_group "inviscid-cp-residual" \
  --inviscid_residual \
  [+ all baseline flags]

CUDA_VISIBLE_DEVICES=1 python train.py --agent frieren --seed 73 \
  --wandb_name "frieren/inviscid-residual-s73" --wandb_group "inviscid-cp-residual" \
  --inviscid_residual \
  [+ all baseline flags]
```

**Important notes:**
- The val metrics MUST be computed on the full pressure (residual + inviscid), not on the residual. Otherwise comparison to baseline is meaningless.
- Start by running a quick debug to verify the inviscid Cp is reasonable (should be -4 to +2 range, with leading-edge peak).
- The existing --residual_prediction flag predicts delta from mean — this stacks with inviscid_residual (the model predicts delta from inviscid Cp, which is already a delta from mean).

## Baseline

- **p_in:** 11.742 | **p_oodc:** 7.643 | **p_tan:** 27.874 | **p_re:** 6.419
- Baseline W&B: k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2324 because ADVISOR: Closing based on W&B results (s70r4u60, i45sp737) — inviscid Cp residual target catastrophically regresses all metrics:

| Metric | 2-seed avg | Baseline | Δ |
|--------|------------|----------|---|
| p_in | 14.815 | 11.742 | +26.2% |
| p_oodc | 12.665 | 7.643 | +65.7% |
| p_tan | 30.465 | 27.874 | +9.3% |
| p_re | 9.500 | 6.419 | +48.0% |

**Root cause:** The thin-airfoil inviscid Cp approximation is too crude for our geometry variety — it assumes flat-plate geometry, ignores NACA thickness/camber, and has leading-edge singularities. When the target becomes p_viscous - p_inviscid, errors in the inviscid estimate propagate directly into the residual target, making it harder to learn rather than easier.

This contrasts with thorfinn's panel-method Cp *feature* approach (which improved p_tan by 3.4%) — using inviscid Cp as an input hint is helpful, but using it to restructure the target is harmful. The model loses the strong inductive bias from learning raw pressure directly.

Moving you to a fresh experiment..

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Inviscid Cp Residual Prediction (classical thin-airfoil formula)

| Metric | Baseline (2-seed avg) | Seed 42 | Seed 73 | 2-seed avg | Δ |
|--------|----------------------|---------|---------|------------|---|
| **p_in** | **11.742** | 14.8 | 14.8 | **14.8** | **+26.1% ❌** |
| **p_tan** | **27.874** | 30.4 | 30.6 | **30.5** | **+9.4% ❌** |
| **p_oodc** | **7.643** | 12.4 | 12.9 | **12.65** | **+65.5% ❌** |
| **p_re** | **6.419** | 9.2 | 9.8 | **9.5** | **+48.0% ❌** |

**Verdict: All metrics significantly regressed. This approach does not work.**

### Run details

- **W&B runs:** s70r4u60 (seed 42), i45sp737 (seed 73)
- **W&B group:** inviscid-cp-residual
- **Epochs:** 144 (both seeds, at 180-min timeout)
- **Peak VRAM:** ~46 GB

### Command used

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/inviscid-residual-s42" --wandb_group "inviscid-cp-residual" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling \
  --inviscid_residual
```

### What happened

The inviscid Cp residual prediction approach catastrophically degrades all surface pressure metrics. The hypothesis was that subtracting an analytical thin-airfoil Cp from the training target would make the model only learn the viscous correction, which should be smaller and easier. In practice, this made performance significantly worse.

**Key findings from research and analysis:**

1. **Formula correction:** The PR's original formula `2*sin(α)/sqrt(ξ*(1-ξ))` was symmetric and singular at both LE and TE, which violates the Kutta condition. I corrected it to the classical thin-airfoil result: `2α*sqrt((1-ξ)/ξ)` which is asymmetric (singular at LE only, zero at TE). However, even with the correct formula the results are poor.

2. **Why it likely fails:**
   - The thin-airfoil approximation is a very crude model for NACA airfoils at moderate-to-high AoA. It doesn't capture thickness effects, camber, or viscous phenomena at all.
   - For tandem foils, the aft foil sees a completely different effective flowfield due to wake interactions — the isolated-foil inviscid Cp is a poor approximation there.
   - The target subtraction creates a bimodal target distribution: volume nodes get zero inviscid Cp while surface nodes get the correction subtracted. This may confuse the model's shared representation.
   - The existing `--residual_prediction` flag already subtracts freestream from targets. Stacking another subtraction (inviscid Cp) may over-correct.

3. **B-GNN reference paper (arXiv:2503.18638):** Research reveals that B-GNN actually feeds inviscid Cp as an **input feature** (not target subtraction) and uses a panel method (not thin-airfoil approximation) for much higher accuracy. The target subtraction approach in this PR is fundamentally different from what B-GNN does.

### Suggested follow-ups

1. **Input feature approach (B-GNN style):** Instead of subtracting inviscid Cp from the target, try adding it as an additional input feature to the model. This preserves the original target distribution and lets the model decide how to use the physics prior. This has stronger published evidence (88% OOD reduction in B-GNN).
2. **Panel method instead of thin-airfoil:** If revisiting this direction, use a panel method (Hess-Smith) instead of thin-airfoil theory. The thin-airfoil approximation is too crude for the range of AoA and geometries in this dataset.
3. **Don't combine with existing residual_prediction:** If retrying target subtraction, try without `--residual_prediction` to avoid double-subtraction of analytical priors.
